### DISCUSSION TITLE
Compile cmd as silent commands?

### DISCUSSION THREAD

**Discussion by eth3lbert** [wbthomason/packer.nvim#270] [Q&A] [answered]
Hello!
I'm trying to lazy-loading with cmd and packer compile cmd as non-silent commands as follows:
```lua
...
-- plugins
use {  "junegunn/fzf.vim",  cmd = { "GFiles", "GFiles?" } }
...
--------------------------------------------------
-- compiled
...
vim.cmd [[command! -nargs=* -range -bang -complete=file GFiles lua require("packer.load")({'fzf.vim'}, { cmd = "GFiles", l1 = <line1>, l2 = <line2>, bang = <q-bang>, args = <q-args> }, _G.packer_plugins)]]
vim.cmd [[command! -nargs=* -range -bang -complete=file GFiles? lua require("packer.load")({'fzf.vim'}, { cmd = "GFiles?", l1 = <line1>, l2 = <line2>, bang = <q-bang>, args = <q-args> }, _G.packer_plugins)]]
...
```
and this causes error:
```
Error in packer_compiled: Vim(lua):E5108: Error executing lua [string ":lua"]:373: Vim(command):E182: Invalid command name
```

We could simply get around this by compiling the cmd as silent command as follows:
```lua
-- compiled
...
vim.cmd [[silent! command -nargs=* -range -bang -complete=file GFiles lua require("packer.load")({'fzf.vim'}, { cmd = "GFiles", l1 = <line1>, l2 = <line2>, bang = <q-bang>, args = <q-args> }, _G.packer_plugins)]]
vim.cmd [[silent! command -nargs=* -range -bang -complete=file GFiles? lua require("packer.load")({'fzf.vim'}, { cmd = "GFiles?", l1 = <line1>, l2 = <line2>, bang = <q-bang>, args = <q-args> }, _G.packer_plugins)]]
...
```

BTW, dein also compile as silent commands as follows:
```
silent! command -complete=customlist,dein#autoload#_dummy_complete -bang -bar -range -nargs=* GFiles call dein#autoload#_on_cmd('GFiles', 'fzf.vim', <q-args>,  expand('<bang>'), expand('<line1>'), expand('<line2>'))
silent! command -complete=customlist,dein#autoload#_dummy_complete -bang -bar -range -nargs=* GFiles? call dein#autoload#_on_cmd('GFiles?', 'fzf.vim', <q-args>,  expand('<bang>'), expand('<line1>'), expand('<line2>'))
```

Any thoughts?

**Comment by wbthomason** [comment, marked as answer]
I'll admit I'm not sure why making the commands `silent` fixes this, but I also don't know of a reason not to make that change.

**Reply by eth3lbert** [reply]
Thanks for the reply, yes, I though vim only support alpha-number as command name, and the silent one seems a little tricky :D
Should I make this small PR or just wait?

**Reply by wbthomason** [reply]
If you have a chance to make a PR, that would be great! Otherwise, I should be able to get to this the next time I do `packer` work (probably in a couple days).

Do you have a pointer to why `silent` fixes this issue, out of curiosity?

**Reply by eth3lbert** [reply]
Nop, I also have no idea.
I just try to migrate my config from dein to packer and encounter this error.
Then I take a look the state file which dein generated and found this one.

**Reply by eth3lbert** [reply]
I just make the pr, and I'll mark this as answered 😄

**Reply by wbthomason** [reply]
Ah. I think I figured out what's going on here. I believe adding `silent!` just masks the error - in the case that we create commands with `silent!`, you don't get a `GFiles?` command, as far as I can tell. Can I ask what that command is meant to be in your config?

**Reply by eth3lbert** [reply]
Actually, I do get the `GFiles?` command, but the error (invalid command name) occurs every time you open the nvim.
```
-- fzf.vim commands
-- See: https://github.com/junegunn/fzf.vim#commands
...
:GFiles? | Git files (git status)
:History: | Command history
:History/ | Search history
...
```
I just copy all fzf.vim commands and put it into `cmd`, and those name with non-alpha-number character `GFiles?`, `History:`, `History/` all cause the invalid command name error.

And `silent!` just bypass/silent the error.

**Reply by wbthomason** [reply]
Right - so, the way those commands are implemented in `fzf.vim` appears to be using the acceptable part of the name (e.g. just `:GFiles`) and then some handling of any characters that come after the name as "arguments".

As such, I'm a little hesitant to add `silent!` to lazy-load commands - I think there's a valid argument for wanting to see these errors, and it doesn't seem that using `silent!` actually enables anything that isn't possible otherwise.

**Reply by eth3lbert** [reply]
Actually, it should be treat as different command, but I do agree with your suggestion and that's why I ask first 😄.
Maybe we could also accept `{"GFiles", silent=true}` as cmd value as follows:
```lua
cmd = {"GFiles", {"GFiles?" silent=true}}
```
What do you think?
